src/storage/src/sink/kafka.rs

```diff
@@ -440,6 +440,27 @@ impl KafkaSinkState {
                     ContextCreationError::Other(anyhow::anyhow!("synthetic error"))
                 ));
 
+                // Create a producer with the old transactional id to fence out writers from older
+                // versions, if any. This section should be removed after it has been running in
+                // production for a week or two.
```

Comment:
Should we be worried about [this sort of thing](https://github.com/MaterializeInc/materialize/issues/23198#issuecomment-1814639317) here?

Reply:
This is indeed a possible outcome but given that we're talking about a one-off event (the release that gets this out) and that during that release the older version will have already been running for a while the probability is low enough that I don't personally worry about it. 

Beyond that this change is taking the system as a whole from an unsafe state (since today this can happen during a cluster resize) to a safe state so I think even more risky moves would be warranted (but I don't consider this one of them).

Did you have in mind additional actions that we could take to minimize the probability further? (assuming this above your risk threshold).

Reply:
> the probability is low enough that I don't personally worry about it.

I guess I don't have a strong enough sense of the odds to feel worried about concurrently running stale sinks (ie. why this code exists) but not worried about those sinks restarting. But if you do, that's great - mostly just checking we'd considered.

> Did you have in mind additional actions that we could take to minimize the probability further?

My typical bias is to try and offload one-off things like this to operations - things like downing the sink-running clusters during the release process in kubernetes - instead of writing code that only gets run once. But we do have a little test coverage of this stuff via the upgrade tests, of course, and manual steps are annoying. But again, if y'all are comfortable with the odds than that seems fine!

Reply:
> of the odds to feel worried about concurrently running stale sinks (ie. why this code exists) but not worried about those sinks restarting.

My reasoning is that from now till eternity cluster resizes will happen many many times whereas this release process will happen once, which is why I'm worried about the former but not the latter. 

> My typical bias is to try and offload one-off things like this to operations - things like downing the sink-running clusters during the release process in kubernetes

Wait, I thought this was the way k8s updated things anyway. Are old pods not being shutdown before the new ones are started? If not then I agree we should coordinate with operations to do this extra step during the release

Reply:
> Wait, I thought this was the way k8s updated things anyway. Are old pods not being shutdown before the new ones are started? If not then I agree we should coordinate with operations to do this extra step during the release

At present, yes—unless there's a node failure, in which case we've configured Kubernetes to assume (rather than verify) that the old pods are dead. I'm comfortable with the level of risk here. It's very unlikely that we'll have a network partition that presents as a node failure at the exact moment of a version upgrade.